materialize-s3-parquet/parquet_file_processor.go

```diff
@@ -0,0 +1,366 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/estuary/protocols/fdb/tuple"
+	pf "github.com/estuary/protocols/flow"
+	pm "github.com/estuary/protocols/materialize"
+	"github.com/xitongsys/parquet-go-source/local"
+
+	"github.com/benbjohnson/clock"
+
+	// TODO revisit this after https://issues.apache.org/jira/browse/ARROW-13986 is completed.
+	"github.com/xitongsys/parquet-go/parquet"
+	"github.com/xitongsys/parquet-go/source"
+	"github.com/xitongsys/parquet-go/writer"
+)
+
+// FileProcessor defines an interface to materialize data into files.
+type FileProcessor interface {
+	// Stores data (key values) to a local file dedicated to the specified binding.
+	Store(binding int, key tuple.Tuple, values tuple.Tuple) error
+
+	// Uploads local files to the cloud storage for all bindings, and returns the updated nextSeqNumList upon success.
+	// Refer to `struct driverCheckpoint` definition in driver.go for details of nextSeqNumList.
+	Commit() (nextSeqNumList []int, e error)
+
+	// Releases resources occupied by the processor.
+	Destroy() error
+}
+
+// FileProcessorProxy implements FileProcessor interface.
+// It proxies a file processor with the following additional logic:
+// a) It maintains a ticker, which triggers the Commit() command on the proxied file processor
+//      if no Commit call to the proxied file processor is observed over a given period of time.
+//      It makes sure that no local file remains in local without being uploaded to the cloud.
+// b) It ensures the mutual exclusive access to the proxied file processor between the ticker
+//      and the transactor(via API calls to the proxy).
+type FileProcessorProxy struct {
+	ctx context.Context
+
+	// The proxied file processor.
+	fileProcessor FileProcessor
+
+	// The time interval, during which if no commit is triggered from the transactor, the proxy triggers one.
+	forceUploadInterval time.Duration
+
+	// A mutex to control mutual exclusive access to the proxied fileProcessor,
+	// and related states of hasLocalStagingData and mostRecentCommitTime
+	mu *sync.Mutex
```

Comment:
A common pattern is to include a `sync.Mutex` as an embedded item in a struct (just omit the `mu` field name). If you have that then you can just call `fp.Lock()` and `fp.Unlock()` on the whole struct. It doesn't hurt to have it as a field but I don't think you want a pointer to it. Since you pass around the FileProcessorProxy as a pointer you don't need the mutex to be a reference also.

Reply:
Done

Reply:
That looks better. You don't actually need to initialize the mutex in your struct. It will be set to the default value which is usable as it is.

Reply:
done